src/control/server/instance_storage_rpc.go

```diff
@@ -185,6 +187,32 @@ func addLinkInfoToHealthStats(prov hardware.PCIeLinkStatsProvider, pciCfg string
 	return nil
 }
 
+func newLinkEvent(id events.RASID, typ, addr, maxVal, negVal string, port uint32) *events.RASEvent {
+	msg := fmt.Sprintf("nvme ssd %q link %s downgraded (port: %d, max: %s, negotiated: %s)",
+		addr, typ, port, maxVal, negVal)
+	return events.NewGenericEvent(id, events.RASSeverityError, msg, "")
+}
+
+func publishLinkStatWarning(engine Engine, pciAddr string, health *ctlpb.BioHealthResp) {
+	if health.LinkMaxSpeed != health.LinkNegSpeed {
+		engine.Debugf("downgraded link speed on %s, publishing ras event", pciAddr)
```

Comment:
This assumes that changes in speed are always downgrades, which I don't think is the case, is it?

In any case, I think I agree that a downgrade probably warrants a higher-priority event than an upgrade. Is it possible to use the same event type with different severities, or is severity an intrinsic property of the event type?

Reply:
The problem here is that we don't store the previous state, i.e. currently there is no way of knowing what the previous health stat was when we do a scan. so we don't know when a transition from degraded to normal happens. It doesn't make sense to report whenever we have a normal state in this case. I will think about how we could store the degraded state.

Reply:
Ah, I see... Hmm. Yeah, and the other issue I think is that couldn't it be normal that the negotiated speed isn't always the max? I don't know for certain, but I would imagine that it's not uncommon for the negotiated speed to be less than the max from this side's perspective. For example, when you have a newer-generation NIC connecting to an older switch.

In that case, maybe it's worth an initial event to indicate that the devices haven't negotiated max speed. I don't think you'd want to raise more events unless something has changed, though.

Reply:
That would require storing the state somewhere though, right? Currently no state is stored.

Reply:
I think you would generally want an event if negotiated speed isn't max, even if that's because of a switch bottleneck.

Reply:
From my side, I agree that it should be relevant to have such events for support tasks: knowing that some interfaces are not working at full speed could be helpful for performance tuning.

Reply:
Implemented: events generated when either transitioning from or to negotiated==maximum state for either speed or width